src/main/scala/beam/sim/vehicles/IncomeBasedVehiclesAdjustment.scala

```diff
@@ -0,0 +1,110 @@
+package beam.sim.vehicles
+
+import beam.agentsim.agents.Population
+import beam.agentsim.agents.vehicles.BeamVehicleType
+import beam.agentsim.agents.vehicles.VehicleCategory.VehicleCategory
+import beam.sim.{BeamScenario, BeamServices}
+import org.apache.commons.math3.distribution.UniformRealDistribution
+import org.matsim.api.core.v01.Coord
+
+case class IncomeBasedVehiclesAdjustment(beamScenario: BeamScenario) extends VehiclesAdjustment {
+
+  private val vehicleTypesAndProbabilityByCategoryAndGroup =
+    scala.collection.mutable.Map[(VehicleCategory, String), Array[(BeamVehicleType, Double)]]()
+  beamScenario.vehicleTypes.values.groupBy(x => x.vehicleCategory).map {
+    case (cat, vehTypes) =>
+      val submap = getCategoryAndGroup(cat, vehTypes.toArray)
+      vehicleTypesAndProbabilityByCategoryAndGroup ++= submap
+  }
+
+  override def sampleVehicleTypesForHousehold(
+    numVehicles: Int,
+    vehicleCategory: VehicleCategory,
+    householdIncome: Double,
+    householdSize: Int,
+    householdPopulation: Population,
+    householdLocation: Coord,
+    realDistribution: UniformRealDistribution,
+  ): List[BeamVehicleType] = {
+    val categoryAndGroup = getHouseholdIncomeGroup(householdIncome)
+    val vehTypeWithProbabilityOption = vehicleTypesAndProbabilityByCategoryAndGroup.get(categoryAndGroup)
+    val vehTypeWithProbability: Array[(BeamVehicleType, Double)] = vehTypeWithProbabilityOption match {
+      case Some(vtWithProb) => vtWithProb
+      case _ =>
+        logger.warn(
+          s"There is no vehicle defined for group ${categoryAndGroup._2}, defaulting to ${vehicleTypesAndProbabilityByCategoryAndGroup.head._1._2}"
+        )
+        vehicleTypesAndProbabilityByCategoryAndGroup.head._2
+    }
+    (1 to numVehicles).map { _ =>
+      val newRand = realDistribution.sample()
+      val (vehType, _) = vehTypeWithProbability.find { case (_, prob) => prob >= newRand }.get
+      vehType
+    }.toList
+  }
+
+  override def sampleRideHailVehicleTypes(
+    numVehicles: Int,
+    vehicleCategory: VehicleCategory,
+    realDistribution: UniformRealDistribution
+  ): List[BeamVehicleType] = {
+    val vehTypeWithProbabilityOption =
+      vehicleTypesAndProbabilityByCategoryAndGroup.get((beam.agentsim.agents.vehicles.VehicleCategory.Car, "ridehail"))
+    val vehTypeWithProbability: Array[(BeamVehicleType, Double)] = vehTypeWithProbabilityOption match {
+      case Some(vtWithProb) => vtWithProb
+      case _ =>
+        logger.warn(
+          s"There is no vehicle defined for ridehail vehicles, defaulting to ${vehicleTypesAndProbabilityByCategoryAndGroup.head._1._2}"
+        )
+        vehicleTypesAndProbabilityByCategoryAndGroup.head._2
+    }
+    (1 to numVehicles).map { _ =>
+      val newRand = realDistribution.sample()
+      val (vehType, _) = vehTypeWithProbability.find { case (_, prob) => prob >= newRand }.get
+      vehType
+    }.toList
+  }
+
+  private def getHouseholdIncomeGroup(
```

Comment:
Why couldn't we use Ranges instead of hard coding? What if someone who is not American wants to use BEAM, do they have to convert their incomes into USD? Why the random dividing points on these ranges? Why not 0 - 20k, 20k - 40k, etc?

I don't want to slow things down seeking perfection, but I'm also worried these issues will require a refactor in the future at a time that is also inconvenient

Reply:
That makes sense. The current oddly-specific ranges are the ones used by NREL, but it wouldn't be too hard to have the ranges themselves defined by the group ID provided in the input file.